### ISSUE TITLE
Add first class support for kotlinx.serialization

### ISSUE THREAD

**Issue by neostage** [spring-projects/spring-boot#24238]
using Spring Boot 2.4.0)
spring-boot-starter-actuator has strong dependencies on jackson so that I can't use kotlinx.serialization as default json message converter.

**Comment by sdeleuze** [comment]
Hi, thanks for raising this since indeed I think there is a need for a proper integration for Kotlin serialization in Boot that goes further than what is inherited from Spring Framework.

Current behavior is [documented here](https://docs.spring.io/spring-framework/docs/current/reference/html/languages.html#kotlin-multiplatform-serialization) and has several limitations in Boot context that we would be likely to improve.

The configuration is kind of Spring Framework oriented, a dedicated Boot support similar to [JSON-B](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/http/JsonbHttpMessageConvertersConfiguration.java) and [related documentation](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#boot-features-json) would be much better for consistency.

The second point that should be improved is what you raised in the issue. After discussing with @bclozel I tend to think it is currently not straightforward to support Kotlin serialization and actuators because the web configuration is shared between your app (that wants to use Kotlin serialization) and and actuator (that needs Jackson or similar general purpose JSON library to be configured in your web configuration. And Kotlin serialization is designed to serialize only Kotlin classes annotated with `@Serializable` or collection of those.

On Framework side, maybe I could refine the converter selection mechanism to have a better support for having both Kotlin Serialization and Jackson via https://github.com/spring-projects/spring-framework/issues/26147 in order to provide a workaround with manual config of both.

On Boot side, after discussing with @snicoll and @bclozel , it sounds like a candidate for Spring Boot 2.5 that would potentially involve the resolution of #20291.

**Comment by neostage** [comment]
Thanks for the detailed response and I'll look forward to future releases!

**Comment by sdeleuze** [comment]
@snicoll After working on a draft commit for spring-projects/spring-framework#26147, I confirm that I plan to configure both Kotlin serialization and Jackson when they are on the classpath, I think that's will make configuration easier and will make actuator and error endpoints still working.

So for Boot 2.5, what would be mainly needed is autoconfiguration and documentation à la JSONB, and dependency management with the upcoming BOM that I have asked to Kotlin team (similar to Coroutines one).

**Comment by snicoll** [comment]
Thanks for the follow-up Sébastien. I've repurposed this issue and triaged it for `2.5.x`.

**Comment by sdeleuze** [comment]
@neostage I have fixed spring-projects/spring-framework#26147 and made sure your use case work fine with Boot 2.4, please test Spring Framework `5.3.2-SNAPSHOT` with both Jackson and Kotlin serialization in the classpath it should work as expected.

**Comment by bclozel** [comment]
I have researched this issue's background, including #39853, spring-projects/spring-framework#26147, spring-projects/spring-framework#32382 and spring-projects/spring-framework#32384.

It seems that for the most part (and partially this initial request https://github.com/spring-projects/spring-boot/issues/24238#issuecomment-734250743), the main problem is that both Jackson and Kotlin Serialization would conflict with each other depending on the available dependencies and the ordering of message converters in Framework's defaults.

I think that spring-projects/spring-framework#26147 already solved quite a bit, by restricting which classes Kotlin Serialization will be involved with. As far as I understand, the Kotlin Serialization converter is now ordered ahead of Jackson by Spring Framework and is only involved if the type to be serialized is annotated with `@kotlinx.serialization.Serializable`. [From this comment](https://github.com/spring-projects/spring-boot/issues/39853#issuecomment-1991240470), I'm not sure we should exclude Kotlin Serialization if 1) it is on the classpath and 2) the class is annotated with `@kotlinx.serialization.Serializable`.

To achieve the initial request:

> spring-boot-starter-actuator has strong dependencies on jackson so that I can't use kotlinx.serialization as default json message converter.

We would need to annotated all actuator types (and possibly exposed library types) with `@kotlinx.serialization.Serializable` and verify that the serialization format is somewhat consistent with what is supported already. This would also apply to the broader Spring Boot ecosystem, since libraries and applications can expose their own.

We could also add a `spring.mvc.converters.preferred-json-mapper=kotlinx`, but this would mean that Jackson would not be registered at all as a converter and that Actuator endpoints would not work unless everything is `@kotlinx.serialization.Serializable`. With this property, wouldn't setting `spring.mvc.converters.preferred-json-mapper=jackson` still would add Kotlin Serialization as a default and ahead of Jackson?

In short, I'm wondering if the main goal of this issue is still current and how these days applications are using Kotlin Serialization/Jackson. @neostage, did spring-projects/spring-framework#26147 solve things for you? How a "Kotlin Serialization without Jackson" situation would work out for you?

**Comment by wilkinsona** [comment]
Thanks for digging into this, Brian.

I think that Actuator using kotlinx.serialization instead of Jackson should be considered separately. It's another variant of https://github.com/spring-projects/spring-boot/issues/13766.

> is only involved if the type to be serialized is annotated with @kotlinx.serialization.Serializable

This is what I thought too but it was never completely clear to me from https://github.com/spring-projects/spring-boot/issues/39853 that it was definitely the case. Reading between the lines a bit, I think the goal was to serialize using Jackson a third-party class that could be serialized using kotlinx. That would require setting the preferred json mapper to Jackson to actually prevent the use of kotlinx. We'd then also want to support setting it to kotlinx to allow folks to opt back out.

>  this would mean that Jackson would not be registered at all as a converter and that Actuator endpoints would not work

I think we're OK here thanks to https://github.com/spring-projects/spring-boot/issues/20291. As long as you have Jackson on the classpath, Actuator would continue to use the Actuator-specific `ObjectMapper` while the application's own endpoints would prefer kotlinx.

**Comment by bclozel** [comment]
Thanks Andy, this one requires quite a bit of background...!

> This is what I thought too but it was never completely clear to me from https://github.com/spring-projects/spring-boot/issues/39853 that it was definitely the case.

Same. I'm wondering if kotlinx still kicks in, in case you're serializing a basic Java type or some Java enum.

> Reading between the lines a bit, I think the goal was to serialize using Jackson a third-party class that could be serialized using kotlinx. That would require setting the preferred json mapper to Jackson to actually prevent the use of kotlinx. We'd then also want to support setting it to kotlinx to allow folks to opt back out.

From what I see, [Spring Framework is adding both Jackson and Kotlinx if present](https://github.com/spring-projects/spring-framework/blob/de2335bee3f5ae57d5a95df727fb471a5a9ba041/spring-webmvc/src/main/java/org/springframework/web/servlet/config/annotation/WebMvcConfigurationSupport.java#L913-L935) (unlike Jackson/Gson/Jsonb where we only pick one).

If I understand correctly, we should add kotlinx to the list of [EQUIVALENT_CONVERTERS](https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/http/HttpMessageConverters.java#L71-L74), meaning that:

* `spring.mvc.converters.preferred-json-mapper=jackson` - kotlinx is completely removed from the picture, only jackson is used
* `spring.mvc.converters.preferred-json-mapper=kotlinx` - Jackson is completely removed from the picture, only kotlinx is used

With `spring.mvc.converters.preferred-json-mapper` matching `jackson` by default, this means that we won't be able to provide the default Spring Framework experience where you get both and kotlinx is checked first and Jackson acts as a fallback? In your comments in #39853, you seem to imply that we should override the ordering from Framework, but from my perspective it's that kotlinx > jackson ordering that allows jackson to be used as a fallback.

I'm probably missing something here...

**Comment by wilkinsona** [comment]
I think it's me that missed something. I'd overlooked the current, and useful, fallback behavior in Framework where kotlinx is tried first with Jackson as a fallback.

> kotlinx is completely removed from the picture, only jackson is used

I think this would be fine.

> Jackson is completely removed from the picture, only kotlinx is used

I think this may be a problem and we'd want it instead to restore Framework's default behavior where it'll use kotlinx if it can and fall back to Jackson as needed.

**Comment by sdeleuze** [comment]
> I think this may be a problem and we'd want it instead to restore Framework's default behavior where it'll use kotlinx if it can and fall back to Jackson as needed.

Indeed if both Jackson and Kotlinx Serialization JSON are present, you likely want to configure both with Jackson as a fallback because Kotlinx Serialization JSON will be used only for class annotated with `@Serializable` processed at build time by the related Kotlin compiler plugin, as well as for related collections.

For other use cases, no Kotlin serializer will be found, and the default behavior should likely be to fallback on Jackson.

**Comment by bclozel** [comment]
We discussed this as a team today and here is our assessment:

Prior to Spring Framework 5.3.2, application that had kotlinx serialization on the classpath would indeed get objects serialized by this library even though they intended to use Jackson. As of Spring Framework 5.3.2 (and https://github.com/spring-projects/spring-framework/issues/26147), this is no longer the case. This serialization library is only involved if:

* kotlinx serialization is on the classpath
* types are annotated with `@kotlinx.serialization.Serializable`

In other cases, Gson or Jackson are being used.

We believe that most cases are now solved thanks to the Spring Framework change.
We are now turning this issue into a documentation improvement to let developers know that if they run into similar issues, they should consider implementing custom DTOs (that are not kotlinx annotated) or manually configure message converters (as described in #39853).

If you are using Spring Framework 5.3.2+ and are still seeing an issue that's not reflected by this comment, please add a comment here with a minimal sample application so we can have a look.

**Comment by hantsy** [comment]
I found when jackson and Kotlin serialization in the classpath, the `WebTestClient` also selects Kotlin serialization over Jackson, even though I have set `spring.http.converters.preferred-json-mapper=jackson`.

**Comment by wilkinsona** [comment]
In that case, the Spring Framework change described above by @bclozel is not working as expected. Please provide a minimal sample that reproduces the behavior you've described.

**Comment by hantsy** [comment]
@wilkinsona The `spring.http.converters.preferred-json-mapper=jackson` from Google result is my fault, I can not find this in the Spring Boot reference. It only includes such a property for `mvc` stack.

I have prepared an example project to describe the issues, 
[webflux-json-mapper.zip](https://github.com/user-attachments/files/17172130/webflux-json-mapper.zip)

I have added `kotlinx-serialization-json` into the dependencies, and then run the `HelloControllerTest`, it will throw an exception like this.

```bash 
Decoding error: Unexpected JSON token at offset 0: Expected beginning of the string, but got [ at path: $
JSON input: ["RED","GREEN","BLUE"]
org.springframework.core.codec.DecodingException: Decoding error: Unexpected JSON token at offset 0: Expected beginning of the string, but got [ at path: $
JSON input: ["RED","GREEN","BLUE"]
	at org.springframework.http.codec.KotlinSerializationStringDecoder.processException(KotlinSerializationStringDecoder.java:139)
	Suppressed: The stacktrace has been enhanced by Reactor, refer to additional information below: 
Error has been observed at the following site(s):
	*__checkpoint ⇢ Body from GET /hello [DefaultClientResponse]
Original Stack Trace:
		at org.springframework.http.codec.KotlinSerializationStringDecoder.processException(KotlinSerializationStringDecoder.java:139)
		at org.springframework.http.codec.KotlinSerializationStringDecoder.lambda$decode$0(KotlinSerializationStringDecoder.java:110)
		at reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onNext(FluxHandleFuseable.java:179)
		at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onNext(FluxMapFuseable.java:299)
		at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onNext(FluxContextWrite.java:107)
		at reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onNext(FluxDoFinally.java:113)
		at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onNext(FluxConcatArray.java:180)
		at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2571)
		at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onSubscribe(FluxConcatArray.java:172)
		at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55)
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:53)
		at reactor.core.publisher.Mono.subscribe(Mono.java:4576)
		at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete(FluxConcatArray.java:238)
		at reactor.core.publisher.FluxConcatArray.subscribe(FluxConcatArray.java:79)
		at reactor.core.publisher.InternalFluxOperator.subscribe(InternalFluxOperator.java:68)
		at reactor.core.publisher.FluxDefer.subscribe(FluxDefer.java:54)
		at reactor.core.publisher.Mono.subscribe(Mono.java:4576)
		at reactor.core.publisher.Mono.block(Mono.java:1806)
		at org.springframework.test.web.reactive.server.DefaultWebTestClient$DefaultResponseSpec.getListBodySpec(DefaultWebTestClient.java:460)
		at org.springframework.test.web.reactive.server.DefaultWebTestClient$DefaultResponseSpec.expectBodyList(DefaultWebTestClient.java:450)
		at com.example.demo.HelloControllerTest.test hello endpoint(HelloControllerTest.kt:20)
		at java.base/java.lang.reflect.Method.invoke(Method.java:580)
		at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
		at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	Suppressed: java.lang.Exception: #block terminated with an error
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:146)
		at reactor.core.publisher.Mono.block(Mono.java:1807)
		at org.springframework.test.web.reactive.server.DefaultWebTestClient$DefaultResponseSpec.getListBodySpec(DefaultWebTestClient.java:460)
		at org.springframework.test.web.reactive.server.DefaultWebTestClient$DefaultResponseSpec.expectBodyList(DefaultWebTestClient.java:450)
		at com.example.demo.HelloControllerTest.test hello endpoint(HelloControllerTest.kt:20)
		at java.base/java.lang.reflect.Method.invoke(Method.java:580)
		at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
		at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 0: Expected beginning of the string, but got [ at path: $
JSON input: ["RED","GREEN","BLUE"]
	at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:24)
	at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:32)
	at kotlinx.serialization.json.internal.AbstractJsonLexer.fail(AbstractJsonLexer.kt:598)
	at kotlinx.serialization.json.internal.AbstractJsonLexer.fail$default(AbstractJsonLexer.kt:596)
	at kotlinx.serialization.json.internal.AbstractJsonLexer.consumeStringLenient(AbstractJsonLexer.kt:467)
	at kotlinx.serialization.json.internal.AbstractJsonLexer.unexpectedToken(AbstractJsonLexer.kt:220)
	at kotlinx.serialization.json.internal.StringJsonLexer.consumeNextToken(StringJsonLexer.kt:74)
	at kotlinx.serialization.json.internal.StringJsonLexer.consumeKeyString(StringJsonLexer.kt:86)
	at kotlinx.serialization.json.internal.AbstractJsonLexer.consumeString(AbstractJsonLexer.kt:383)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeString(StreamingJsonDecoder.kt:339)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeEnum(StreamingJsonDecoder.kt:352)
	at kotlinx.serialization.internal.EnumSerializer.deserialize(Enums.kt:139)
	at kotlinx.serialization.internal.EnumSerializer.deserialize(Enums.kt:105)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:69)
	at kotlinx.serialization.json.Json.decodeFromString(Json.kt:107)
	at org.springframework.http.codec.KotlinSerializationStringDecoder.lambda$decode$0(KotlinSerializationStringDecoder.java:107)
	at reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onNext(FluxHandleFuseable.java:179)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onNext(FluxMapFuseable.java:299)
	at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onNext(FluxContextWrite.java:107)
	at reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onNext(FluxDoFinally.java:113)
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onNext(FluxConcatArray.java:180)
	at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2571)
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onSubscribe(FluxConcatArray.java:172)
	at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55)
	at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:53)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4576)
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete(FluxConcatArray.java:238)
	at reactor.core.publisher.FluxConcatArray.subscribe(FluxConcatArray.java:79)
	at reactor.core.publisher.InternalFluxOperator.subscribe(InternalFluxOperator.java:68)
	at reactor.core.publisher.FluxDefer.subscribe(FluxDefer.java:54)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4576)
	at reactor.core.publisher.Mono.block(Mono.java:1806)
	at org.springframework.test.web.reactive.server.DefaultWebTestClient$DefaultResponseSpec.getListBodySpec(DefaultWebTestClient.java:460)
	at org.springframework.test.web.reactive.server.DefaultWebTestClient$DefaultResponseSpec.expectBodyList(DefaultWebTestClient.java:450)
	at com.example.demo.HelloControllerTest.test hello endpoint(HelloControllerTest.kt:20)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)


Unexpected JSON token at offset 0: Expected beginning of the string, but got [ at path: $
JSON input: ["RED","GREEN","BLUE"]
kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 0: Expected beginning of the string, but got [ at path: $
JSON input: ["RED","GREEN","BLUE"]
	at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:24)
	at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:32)
	at kotlinx.serialization.json.internal.AbstractJsonLexer.fail(AbstractJsonLexer.kt:598)
	at kotlinx.serialization.json.internal.AbstractJsonLexer.fail$default(AbstractJsonLexer.kt:596)
	at kotlinx.serialization.json.internal.AbstractJsonLexer.consumeStringLenient(AbstractJsonLexer.kt:467)
	at kotlinx.serialization.json.internal.AbstractJsonLexer.unexpectedToken(AbstractJsonLexer.kt:220)
	at kotlinx.serialization.json.internal.StringJsonLexer.consumeNextToken(StringJsonLexer.kt:74)
	at kotlinx.serialization.json.internal.StringJsonLexer.consumeKeyString(StringJsonLexer.kt:86)
	at kotlinx.serialization.json.internal.AbstractJsonLexer.consumeString(AbstractJsonLexer.kt:383)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeString(StreamingJsonDecoder.kt:339)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeEnum(StreamingJsonDecoder.kt:352)
	at kotlinx.serialization.internal.EnumSerializer.deserialize(Enums.kt:139)
	at kotlinx.serialization.internal.EnumSerializer.deserialize(Enums.kt:105)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:69)
	at kotlinx.serialization.json.Json.decodeFromString(Json.kt:107)
	at org.springframework.http.codec.KotlinSerializationStringDecoder.lambda$decode$0(KotlinSerializationStringDecoder.java:107)
	at reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onNext(FluxHandleFuseable.java:179)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableConditionalSubscriber.onNext(FluxMapFuseable.java:299)
	at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onNext(FluxContextWrite.java:107)
	at reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onNext(FluxDoFinally.java:113)
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onNext(FluxConcatArray.java:180)
	at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2571)
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onSubscribe(FluxConcatArray.java:172)
	at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55)
	at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:53)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4576)
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete(FluxConcatArray.java:238)
	at reactor.core.publisher.FluxConcatArray.subscribe(FluxConcatArray.java:79)
	at reactor.core.publisher.InternalFluxOperator.subscribe(InternalFluxOperator.java:68)
	at reactor.core.publisher.FluxDefer.subscribe(FluxDefer.java:54)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4576)
	at reactor.core.publisher.Mono.block(Mono.java:1806)
	at org.springframework.test.web.reactive.server.DefaultWebTestClient$DefaultResponseSpec.getListBodySpec(DefaultWebTestClient.java:460)
	at org.springframework.test.web.reactive.server.DefaultWebTestClient$DefaultResponseSpec.expectBodyList(DefaultWebTestClient.java:450)
	at com.example.demo.HelloControllerTest.test hello endpoint(HelloControllerTest.kt:20)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)



```

**Comment by hantsy** [comment]
I have added a custom `CodecCustomizer` to resolve the issue. https://github.com/hantsy/spring-puzzles/blob/master/webflux-json-mapper/src/main/kotlin/com/example/demo/DemoApplication.kt#L41

**Comment by hantsy** [comment]
@bclozel @wilkinsona I created another `WebMvc` example project, https://github.com/hantsy/spring-puzzles/tree/master/webmvc-json-mapper, just expose a enum values in the endpoint, no Kotlinx specific annotation.
* Add kotlinx-serializaiton-json into dependencies
* Add property `spring.mvc.converters.preferred-json-mapper=jackson`

In the [Controller test](https://github.com/hantsy/spring-puzzles/blob/master/webmvc-json-mapper/src/test/kotlin/com/example/demo/HelloControllerTest.kt),  
* When using `MockMvc`, it works well.
* When using `WebTestClient` by `@Autowired` or `MockMvcWebTestClient.bindTo(mockMvc)`, it failed with the above exception, I have to add custom codecs to switch to use Jackson and make it work.

**Comment by hantsy** [comment]
Jackson, Gson and Jsonb have their own `Autoconfiguraiton` classes, and I would like an autoconfiguration for `kotlinx.serialization`. Thus I can exclude the kotlinx serialization auto configuration in a  spring `WebFlux` application.

**Comment by bclozel** [comment]
I guess enum values don't need to be annotated with `@kotlinx.serialization.Serializable` to be considered by kotlinx.serialization. That's a limitation we could document here or in Spring Framework. As for introducing a brand new, dedicated auto-configuration for kotlinx, that's something we can do but I don't think it's high priority given that the only limitation so far is for applications 1) having the dependency on the classpath and 2) serializing enums.

**Comment by hantsy** [comment]
@bclozel In our project, the Kotlinx serialization JSON is a transitive dependency from other libs we used in the project, which we did not add explicitly. We can not control this case.

**Comment by wilkinsona** [comment]
It’s starting to sound like a way to disable kotlinx is required, perhaps through an annotation that Framework’s codecs and message converters look for or through some mechanism in Boot. It would appear that the classpath alone isn’t a sufficiently strong signal.

**Comment by y-marion** [comment]
We also have a similar issue where we have the dependency on `kotest-assertions-json` in `testImplementation` (which brings in `kotlinx.serialization` as transitive dependency).
This causes an error in a `MockMVC` test, where an Endpoint returning `ResponseEntity<Unit>` is called, if we return a ProblemDetail (or any response with a body in case of a failure), as `kotlinx.serialization.internal.UnitSerializer.serialize` is called either way to serialize the Response.

I've written a small reproduction here: https://github.com/y-marion/spring-kotlinx-unit-error-repro
The tests for the controller do not work if executed as is, if the `testImplementation("io.kotest:kotest-assertions-json:5.9.1")` is removed, they work.
The controller also works as intended if the application is started via bootRun.

**Comment by sdeleuze** [comment]
> It’s starting to sound like a way to disable kotlinx is required, perhaps through an annotation that Framework’s codecs and message converters look for or through some mechanism in Boot. It would appear that the classpath alone isn’t a sufficiently strong signal.

@wilkinsona @bclozel Feel free to ping me if you want to brainstorm on what we could do and when.

**Comment by philwebb** [comment]
We're going to add auto-configuration and look at adding an entry to the `spring.mvc.converters.preferred-json-mapper` property.

**Comment by bclozel** [comment]
I did some work on this issue and came to the following conclusion: adding "first class" support for kotlinx.serialization would only be useful if developers want to contribute a custom `kotlinx.serialization.json.Json` bean and configure it to their liking. This is not what has been requested.

Quite the opposite, it seems that kotlinx serialization conflicts with Jackson when they're both on the classpath. While there was an initial attempt to make the kotlin variant more selective (by looking at annotations), this obviously does not work for Java enums. Switching the order won't improve things: if Kotlin is first, it will handle java enums and `@kotlinx.serialization.Serializable`-classes, if Jackson is first, it will handle everything.

The only possible improvement in Spring Boot would be remove the Kotlin converter if `spring.mvc.converters.preferred-json-mapper=jackson`. Still, the default Spring Framework experience is flawed because ordered converters are conflicting with each other. Spring Framework currenty considers Jackson/Gson/Jsonb in this order, configuring a single one to avoid conflicts. I think this should be the same for kotlinx serialization.

I'm closing this issue in favor of https://github.com/spring-projects/spring-framework/issues/34410
